In [1]:
# Initializaion
import turicreate as tc
products = tc.SFrame('../../data/amazon_baby.sframe/')
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

# Use .apply() to build a new feature with the counts for each of the selected_words

In [92]:
# Add word_count column
products['word_count'] = tc.text_analytics.count_words(products['review'])
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'handles': 1.0, 'strippin... |
| {'recommend': 1.0, 'disapp... |
| {'quilt': 1.0, 'the': 1.0,... |
| {'tool': 1.0, 'clever': 1.... |
| {'rock': 1.0, 'many': 1.0,... |
| {'thumb': 1.0, 'or': 1.0, ... |
| {'for': 1.0, 'barnes': 1.0... |
| {'right': 1.0, 'because': ... |
| {'like': 1.0, 'and': 1.0, ... |
| {'in': 1.0, 'pages': 1.0, ... |
+-------------------------------+
[183531 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [93]:
# Define word_counter function
def word_counter(word_count, word):
    return word_count.get(word, 0)

In [94]:
# Add respective columns to the data
for word in selected_words:
    products[word] = products['word_count'].apply(lambda wc: word_counter(wc, word))
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'handles': 1.0, 'strippin... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'recommend': 1.0, 'disapp... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'quilt': 1.0, 'the': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'tool': 1.0, 'clever': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
| {'rock': 1.0, 'many': 1.0,... |   0.0   |  1.0  |    0.0    |   0.0   | 1.0  |
| {'thumb': 1.0, 'or': 1.0, ... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'for': 1.0, 'barnes': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'right': 1.0, 'because': ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'like': 1.0, 'and': 1.0, ... |   0.0   |  0.0  |    1.0    |   0.0   | 0.0  |
| {'in': 1.0, 'pages': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+
| horrible | bad | terrible | awful | wow | hate |
+----------+-----+----------+-------+-----+------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
+----------+-----+----------+-------+-----+------+
[183531 rows x 15 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [95]:
# Sum how much each word in selected_words got mentioned
word_sums = {word: products[word].sum() for word in selected_words}
print(word_sums)

{'awesome': 4075.0, 'great': 59536.0, 'fantastic': 1765.0, 'amazing': 2726.0, 'love': 43867.0, 'horrible': 1245, 'bad': 4950.0, 'terrible': 1282, 'awful': 753, 'wow': 461, 'hate': 1285}


In [96]:
# The most used word in the dataset & the least used one from selected_words
most_used = max(word_sums, key=word_sums.get)
least_used = min(word_sums, key=word_sums.get)
print ('most used: ', most_used, word_sums[most_used])
print ('least used: ', least_used, word_sums[least_used])

most used:  great 59536.0
least used:  wow 461


# Create a new sentiment analysis model using only the selected_words as features

In [97]:
# ignoring neutral reviews
products = products[products['rating']!= 3]
# sentiment = 1 if rating >= 4 otherwise sentiment = 0
products['sentiment'] = products['rating'] >= 4
# train/test split
train_data,test_data = products.random_split(.8, seed=0)
# training the model on selected_words
selected_words_model = tc.logistic_classifier.create(train_data,target='sentiment', features=selected_words, validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients      : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.018917     | 0.847401          | 0.845874            |

| 2         | 3        | 0.035655     | 0.847514          | 0.846085            |

| 3         | 4        | 0.050746     | 0.847626          | 0.846115            |

| 4         | 5        | 0.066754     | 0.847708          | 0.846385            |

| 5         | 6        | 0.083384     | 0.847708          | 0.846385            |

| 6         | 7        | 0.102512     | 0.847708          | 0.846385            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [98]:
# Coefficients
coefficients = selected_words_model.coefficients.sort(key_column_names='value')
coefficients.print_rows(num_rows=len(selected_words) + 1)
# The most positive weight
print ('the most positive weight: ', coefficients[-1]['name'])
# The most negative weight
print ('the most negative weight: ', coefficients[0]['name'])

+-------------+-------+-------+-----------------------+----------------------+
|     name    | index | class |         value         |        stderr        |
+-------------+-------+-------+-----------------------+----------------------+
|   horrible  |  None |   1   |  -2.2513352367590964  | 0.08020249388788438  |
|   terrible  |  None |   1   |  -2.2236614360851306  | 0.07731736203785755  |
|    awful    |  None |   1   |   -2.052908204031355  | 0.10099735435259258  |
|     hate    |  None |   1   |   -1.348440722246315  | 0.07715698604297333  |
|     bad     |  None |   1   |  -0.9914778800650603  | 0.03848428664699064  |
|     wow     |  None |   1   | -0.009538236067682589 | 0.16046411224711654  |
|    great    |  None |   1   |   0.8630655001196608  | 0.018955052444376966 |
|  fantastic  |  None |   1   |   0.8858047568814255  |  0.1116759129339967  |
|   amazing   |  None |   1   |   1.1000933113660254  | 0.09954776260465989  |
|   awesome   |  None |   1   |    1.13353466603414 

# Comparing the accuracy of different sentiment analysis model

In [99]:
# Count how many of each label (1,0) you have
label_counts = products['sentiment'].value_counts()
# Majority baseline accuracy
majority_accuracy = label_counts['count'].max() / label_counts['count'].sum()
print('accuracy of the majority class classifier: ', majority_accuracy)

accuracy of the majority class classifier:  0.8411233448474381


### Importing sentiment_model from FND03-NB01.ipynb

In [100]:
sentiment_model = tc.logistic_classifier.create(train_data,target='sentiment', features=['word_count'], validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 57356

Number of coefficients      : 57357

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 4        | 0.250000  | 0.297940     | 0.841421          | 0.840019            |

| 1         | 9        | 3.250000  | 0.761720     | 0.931359          | 0.911362            |

| 2         | 11       | 2.669978  | 1.011861     | 0.938650          | 0.916466            |

| 3         | 12       | 2.669978  | 1.191587     | 0.927417          | 0.901814            |

| 4         | 14       | 1.326072  | 1.425572     | 0.945200          | 0.918088            |

| 9         | 20       | 1.326072  | 2.292184     | 0.977654          | 0.917698            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

### Comparison

In [101]:
# Evaluating the models
print ('sentiment_model: ', sentiment_model.evaluate(test_data, metric='accuracy'))
print ('selected_words_model: ', selected_words_model.evaluate(test_data, metric='accuracy'))

sentiment_model:  {'accuracy': 0.9176975738650012}
selected_words_model:  {'accuracy': 0.8463848186404036}


selected_words_model’s accuracy is 0.846, just barely better than the dumb baseline of 0.84

While the orginal sentiment_model that was trained on all words was 0.917, way better than the baseline

 # Interpreting the difference in performance between the models

In [107]:
# add predicted_sentiment column
products['predicted_sentiment'] = sentiment_model.predict(products, output_type = 'probability')
# select a particular product
diaper_champ_reviews = products[products['name']== 'Baby Trend Diaper Champ']
# sort the product reviews according to predicted_sentiment
diaper_champ_reviews = diaper_champ_reviews.sort(key_column_names='predicted_sentiment', ascending=False)
# the most positive review for ‘Baby Trend Diaper Champ’, according to the sentiment_model 
most_positive_review = diaper_champ_reviews[0]
print (most_positive_review)

{'name': 'Baby Trend Diaper Champ', 'review': "I read a review below that can explain exactly what we experienced. We've had it for 16 months and it has worked wonderful for us. No smells, change it out once a week, easy to clean. Then a diaper snagged this foam material in the head part, so I pulled the rest of the foam out. Big mistake!!! Now it can no loner retain the stinkiness and we're looking for a replacement. Be careful of overloading and never take out that foam piece that is cushioned between pieces. I have figured out that it is key to keeping the stink out.", 'rating': 4.0, 'word_count': {'key': 1.0, 'have': 1.0, 'pieces': 1.0, 'between': 1.0, 'cushioned': 1.0, 'piece': 1.0, 'take': 1.0, 'overloading': 1.0, 'be': 1.0, 'looking': 1.0, 're': 1.0, 'stinkiness': 1.0, 'retain': 1.0, 'now': 1.0, 'wonderful': 1.0, 'worked': 1.0, '16': 1.0, 'and': 3.0, 'months': 1.0, 've': 1.0, 'in': 1.0, 'us': 1.0, 'i': 3.0, 'experienced': 1.0, 'read': 1.0, 'easy': 1.0, 'for': 3.0, 'to': 2.0, 'ha

In [109]:
# predicted_sentiment for that review, if we use the selected_words_model to analyze it
selected_words_model.predict(diaper_champ_reviews[0:1], output_type='probability')

dtype: float
Rows: 1
[0.7919288370624462]

The value of the predicted_sentiment for the most positive review found using the sentiment_model is much more positive than the value predicted using the selected_words_model because none of the selected_words appeared in the text of this review.